File: ClientInterface.hs
module ClientInterface where

import System.IO (withFile, IOMode(..), hGetContents)
import Data.List (isPrefixOf, find)
import SessionManagement (Session(..)) -- Importa o tipo Session do módulo SessionManagement
import CinemaInfo (viewCinemaInfo)     -- Importa a função viewCinemaInfo do módulo CinemaInfo
import ReviewManagement (leaveReview)  -- Importa a função leaveReview do módulo ReviewManagement

-- Função para visualizar filmes (mantida)
viewMovies :: IO ()
viewMovies = do
  withFile "movies.txt" ReadMode $ \handle -> do
    contents <- hGetContents handle
    putStrLn "Lista de filmes disponíveis:"
    putStrLn contents

-- Função para visualizar itens da bomboniere (mantida)
viewItems :: IO [String]
viewItems = do
  withFile "concessionStand.txt" ReadMode $ \handle -> do
    contents <- hGetContents handle
    let items = lines contents
    putStrLn "Itens da bomboniere disponíveis:"
    if null items
      then putStrLn "Nenhum item disponível." >> return []
      else do
        mapM_ putStrLn items
        return items

-- Função para comprar um item da bomboniere (mantida)
buyItem :: IO ()
buyItem = do
  items <- viewItems
  if null items
    then putStrLn "Voltando ao menu principal..." -- Caso não haja itens, volta ao menu principal
    else do
      putStrLn "Digite o nome do item que deseja comprar:"
      itemName <- getLine
      -- Verifica se o item existe na lista de itens disponíveis
      case find (isPrefixOf itemName) items of
        Just _  -> putStrLn ("Você comprou o item: " ++ itemName)
        Nothing -> putStrLn "Item não encontrado. Voltando ao menu principal."

-- Função para visualizar sessões de cinema
viewSessions :: IO ()
viewSessions = do
  withFile "sessions.txt" ReadMode $ \handle -> do
    contents <- hGetContents handle
    let sessions = read contents :: [Session]
    putStrLn "Sessões disponíveis:"
    if null sessions
      then putStrLn "Nenhuma sessão disponível."
      else mapM_ printSessionDetails sessions

-- Função auxiliar para imprimir os detalhes de uma sessão
printSessionDetails :: Session -> IO ()
printSessionDetails (Session title time room date price) = do
  putStrLn $ "Filme: " ++ title
  putStrLn $ "Data: " ++ date
  putStrLn $ "Horário: " ++ time
  putStrLn $ "Sala: " ++ room
  putStrLn $ "Preço do Ingresso: R$ " ++ show price
  putStrLn "------------------------"

-- Função para comprar um ingresso (atualizada)
buyTicket :: IO ()
buyTicket = do
  putStrLn "Digite o título do filme para o qual deseja comprar o ingresso:"
  movieTitle <- getLine
  putStrLn "Digite a data da sessão (DD/MM):"
  sessionDate <- getLine
  putStrLn "Digite o horário da sessão (HH:MM):"
  sessionTime <- getLine
  
  withFile "sessions.txt" ReadMode $ \handle -> do
    contents <- hGetContents handle
    let sessions = read contents :: [Session]
        foundSession = find (\(Session title time _ date _) -> 
                              title == movieTitle && date == sessionDate && time == sessionTime) sessions

    case foundSession of
      Just (Session title time room date price) -> do
        putStrLn $ "Sessão encontrada!"
        putStrLn $ "Filme: " ++ title
        putStrLn $ "Data: " ++ date
        putStrLn $ "Horário: " ++ time
        putStrLn $ "Sala: " ++ room
        putStrLn $ "Preço do Ingresso: R$ " ++ show price
        
        putStrLn "Você é estudante e tem carteirinha? (s/n):"
        isStudent <- getLine
        let finalPrice = if isStudent == "s" then price * 0.5 else price
        putStrLn $ "Preço final do Ingresso: R$ " ++ show finalPrice

        putStrLn "Digite 's' para confirmar a compra ou qualquer outra tecla para cancelar:"
        confirmation <- getLine
        if confirmation == "s"
          then putStrLn "Ingresso comprado com sucesso!"
          else putStrLn "Compra cancelada."
      Nothing -> do
        putStrLn "Sessão não encontrada."
        putStrLn "1) Tentar novamente"
        putStrLn "2) Voltar ao menu anterior"
        option <- getLine
        case option of
          "1" -> buyTicket
          "2" -> return () -- Volta ao menu anterior
          _   -> putStrLn "Opção inválida. Voltando ao menu anterior."

-- Função para deixar feedback
giveFeedback :: IO ()
giveFeedback = leaveReview

-- Menu do cliente (atualizado)
runClientMode :: IO ()
runClientMode = do
  putStrLn "Modo Cliente:"
  putStrLn "1) Visualizar Filmes"
  putStrLn "2) Visualizar Sessões Disponíveis"
  putStrLn "3) Comprar Ingresso"
  putStrLn "4) Comprar Item da Bomboniere"
  putStrLn "5) Visualizar Informações do Cinema"
  putStrLn "6) Deixar Feedback"
  putStrLn "7) Voltar ao Menu Principal"
  option <- getLine
  case option of
    "1" -> viewMovies >> runClientMode
    "2" -> viewSessions >> runClientMode
    "3" -> buyTicket >> runClientMode
    "4" -> buyItem >> runClientMode
    "5" -> viewCinemaInfo >> runClientMode
    "6" -> giveFeedback >> runClientMode
    "7" -> return () -- Retorna ao menu principal
    _   -> putStrLn "Opção inválida. Tente novamente." >> runClientMode

File: LOGICO(Prolog)/ClientInterface.pl
:- use_module(library(readutil)).

% Importando os modulos
:- consult('SessionManagement').
:- consult('CinemaInfo').
:- consult('ReviewManagement').
:- consult('FAQ').

% Função para visualizar filmes
view_movies :-
    write("Lista de filmes disponiveis:"), nl,
    read_file('movies.txt', Contents),
    write(Contents), nl.

% Função para visualizar lançamentos futuros
view_upcoming_movies :-
    write("Lista de lancamentos futuros:"), nl,
    read_file('upcoming_movies.txt', Contents),
    write(Contents), nl.

% Função para visualizar itens da bomboniere
view_items(Items) :-
    read_file_lines('concessionStand.txt', Items),
    write("Itens da bomboniere disponiveis:"), nl,
    ( Items = [] ->
        write("Nenhum item disponivel."), nl
    ; maplist(write_item, Items)
    ).

write_item(Item) :-
    write(Item), nl.

% Função para comprar um item da bomboniere
buy_item :-
    view_items(Items),
    ( Items = [] ->
        write("Voltando ao menu principal..."), nl
    ; write("Digite o nome do item que deseja comprar:"), nl,
      read_line_to_string(user_input, ItemName),
      ( member(ItemName, Items) ->
          write("Voce comprou o item: "), write(ItemName), nl
      ; write("Item nao encontrado. Voltando ao menu principal."), nl
      )
    ).

% Função para visualizar sessões de cinema
view_sessions :-
    write("Sessoes disponiveis:"), nl,
    read_file('sessions.txt', Contents),
    ( Contents = [] ->
        write("Nenhuma sessao disponivel."), nl
    ; read_term_from_atom(Contents, Sessions, []),
      maplist(print_session_details, Sessions)
    ).

% Função auxiliar para imprimir os detalhes de uma sessão
print_session_details(Session) :-
    Session = session(Title, Time, Room, Date, Price, Audio),
    write("Filme: "), write(Title), nl,
    write("Data: "), write(Date), nl,
    write("Horario: "), write(Time), nl,
    write("Sala: "), write(Room), nl,
    write("Preco do Ingresso: R$ "), write(Price), nl,
    write("Tipo de audio: "), write(Audio), nl,
    write("------------------------"), nl.

% Função para comprar um ingresso
buy_ticket :-
    write("Digite o titulo do filme para o qual deseja comprar o ingresso:"), nl,
    read_line_to_string(user_input, MovieTitle),
    write("Digite a data da sessao (DD/MM):"), nl,
    read_line_to_string(user_input, SessionDate),
    write("Digite o horario da sessao (HH:MM):"), nl,
    read_line_to_string(user_input, SessionTime),
    
    read_file('sessions.txt', Contents),
    read_term_from_atom(Contents, Sessions, []),
    ( member(session(MovieTitle, SessionTime, _, SessionDate, Price, Audio), Sessions) ->
        write("Sessao encontrada!"), nl,
        write("Filme: "), write(MovieTitle), nl,
        write("Data: "), write(SessionDate), nl,
        write("Horario: "), write(SessionTime), nl,
        write("Preco do Ingresso: R$ "), write(Price), nl,
        write("Tipo de audio: "), write(Audio), nl,
        write("Voce eh estudante e tem carteirinha? (s/n):"), nl,
        read_line_to_string(user_input, IsStudent),
        ( IsStudent = "s" ->
            FinalPrice is Price * 0.5
        ; FinalPrice is Price
        ),
        write("Preco final do Ingresso: R$ "), write(FinalPrice), nl,
        write("Digite 's' para confirmar a compra ou qualquer outra tecla para cancelar:"), nl,
        read_line_to_string(user_input, Confirmation),
        ( Confirmation = "s" ->
            write("Ingresso comprado com sucesso!"), nl
        ; write("Compra cancelada."), nl
        )
    ; write("Sessao nao encontrada."), nl,
      write("1) Tentar novamente"), nl,
      write("2) Voltar ao menu anterior"), nl,
      read_line_to_string(user_input, Option),
      ( Option = "1" -> buy_ticket
      ; true
      )
    ).

% Função para deixar feedback
give_feedback :-
    leave_review.

% Função auxiliar para ler o conteúdo de um arquivo
read_file(File, Contents) :-
    open(File, read, Stream),
    read_string(Stream, _, Contents),
    close(Stream).

% Função auxiliar para ler linhas de um arquivo
read_file_lines(File, Lines) :-
    open(File, read, Stream),
    read_lines(Stream, Lines),
    close(Stream).

% Menu do cliente
run_client_mode :-
    write("Modo Cliente:"), nl,
    write("1) Visualizar Filmes"), nl,
    write("2) Visualizar Sessoes Disponiveis"), nl,
    write("3) Visualizar Lancamentos Futuros"), nl,
    write("4) Comprar Ingresso"), nl,
    write("5) Comprar Item da Bomboniere"), nl,
    write("6) Visualizar Informacoes do Cinema"), nl,
    write("7) Deixar Feedback"), nl,
    write("8) Acessar FAQ"), nl,
    write("9) Voltar ao Menu Principal"), nl,

    read_line_to_string(user_input, Option),
    ( Option = "1" -> view_movies, run_client_mode
    ; Option = "2" -> view_sessions, run_client_mode
    ; Option = "3" -> view_upcoming_movies, run_client_mode
    ; Option = "4" -> buy_ticket, run_client_mode
    ; Option = "5" -> buy_item, run_client_mode
    ; Option = "6" -> view_cinema_info, run_client_mode
    ; Option = "7" -> give_feedback, run_client_mode
    ; Option = "8" -> view_faq, run_client_mode
    ; Option = "9" -> true
    ; write("Opcao invalida. Tente novamente."), nl, run_client_mode
    ).
